gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+
+        match value as u16 {
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING => config.set_edge_detection(Edge::Rising),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING => config.set_edge_detection(Edge::Falling),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH => config.set_edge_detection(Edge::Both),
+
+            // Drop the buffer.
+            VIRTIO_GPIO_IRQ_TYPE_NONE => {
+                state.buffer = None;
+                return Ok(());
+            }
+
+            // Only edge IRQs are supported for now.
+            _ => return Err(Error::GpioIrqTypeNotSupported(value)),
+        };
+
+        // Allocate the buffer and configure the line for interrupt.
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
+
+            state
+                .request
+                .as_ref()
+                .unwrap()
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)
+        }
+    }
+
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()> {
+        // While waiting here for the interrupt to occur, it is possible that we receive another
+        // request to disable the interrupt instead, via a call to set_irq_type().
+        //
+        // That request shouldn't wait for this wait_for_interrupt() function to finish, else we
+        // may never proceed with set_irq_type() as the parent wait_for_interrupt() function will
+        // take the lock right after dropping one and hence set_irq_type() will wait for ever.
+        //
+        // To avoid this situation, implement the state.request as an "Arc" instance instead and
+        // take a reference to it without locking the state. With this, while we are waiting for
+        // the interrupt, set_irq_type() can come and set the state.request to None and we will be
+        // able to quit the wait loop.
+        let request = {
+            let state = &self.state[gpio as usize].write().unwrap();
+
+            match &state.request {
+                Some(x) => x.clone(),
+                None => return Err(Error::GpioIrqNotEnabled),
+            }
+        };
+
+        // Wait for the interrupt for a second.
+        match request.edge_event_wait(Duration::new(1, 0)) {
+            Err(LibGpiodError::OperationTimedOut) => return Err(Error::GpioIrqOpTimedOut),
+            x => x.map_err(Error::GpiodFailed)?,
+        }
+
+        // The interrupt has already occurred, we can lock now just fine.
+        let state = &self.state[gpio as usize].write().unwrap();
+        if let Some(buffer) = &state.buffer {
+            request
+                .edge_event_read(buffer, 1)
+                .map_err(Error::GpiodFailed)?;
+
+            Ok(())
+        } else {
+            Err(Error::GpioLineRequestNotConfigured)
+        }
+    }
```

Comment:
The EdgeEvenBuffer of a line state is read, which for understandable reasons, has to be done with the RwLock held.  But it doesn't explain why the reference count on the LineRequest needs to be incremented - at least it is not clear to me.

Reply:
To make sure the `request` instance doesn't get freed while we are waiting for the interrupt. Here is a sequence of events on how it can happen:

- Controller::wait_for_interrupt(); // Waits for a second for the interrupt
   - request.edge_event_wait()

in the meanwhile, this happens:
- Controller::set_irq_type(TYPE_NONE); //disables the interrupt by setting buffer to NONE;
- Controller::set_direction(DIR_NONE); 
   - state.request = NONE; // If we don't take a reference above, request structure will be dropped while still being used.

Reply:
In the original comments found in wait_for_interrupt() and the new one provided above, the call to Controller::set_direction(), althought very important, is never mentionned.  Using an asynchronous reference count will indeed prevent state.request from being free'd when it is set to None in set_direction(), but the exact time that is happening is up to the Rust infrastructure and out of our control.

Moreover, from reading the code and the explanations you are giving me I infer that for edge_event_wait() to work properly both state.request _and_  state.buffer need to be valid.  If that is the case then it is definitely worth mentionning in comments.  

Another problem seems to be the concurrency between the Controller::wait_for_interrupt() loop and other Controller commands such as set_irq_type() and set_direction().  

With the above in mind I think simply yielding the processor (thread::yield_now) right after the match statement that handles device.wait_for_interrupt() would give other operations a chance to acquire the RwLock.  That way it would be possible to:

1) Keep the lock in device.wait_for_interrupt() without fear of starving other requests.
2) Get ride of the Arc type.
3) keep both state.request and state.buffer valid for the duration of the device.wait_for_interrupt().

If that is not acceptable I would simply return a busy error in any operation that can alter state.request and state.buffer  when RwLock is already taken.

Reply:
Thanks for the detailed explanation, it really helps putting the pieces together.  Let's keep the current design and I will review it again in the next revision, along with the updated comment section.

Reply:
Thanks.